File: app/src/main/java/com/hover/stax/views/StaxTextInputLayout.java
package com.hover.stax.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.LayoutInflater;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.hover.stax.R;
import com.hover.stax.databinding.StaxInputBinding;

public class StaxTextInputLayout extends AbstractStatefulInput {
    private final String TAG = "StaxTextInputLayout";

    private String hint;
    private int inputType;
    private TextInputEditText editText;

    StaxInputBinding binding;

    public StaxTextInputLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        getAttrs(context, attrs);

        binding = StaxInputBinding.inflate(LayoutInflater.from(context), this, true);

        initView();
        fillAttr();
    }

    protected void initView() {
        super.initView();
        editText = binding.inputEditText;
    }

    private void getAttrs(Context context, AttributeSet attrs) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.StaxTextInputLayout, 0, 0);
        try {
            hint = a.getString(R.styleable.StaxTextInputLayout_android_hint);
            inputType = a.getInt(R.styleable.StaxTextInputLayout_android_inputType, 0);
        } finally {
            a.recycle();
        }
    }

    private void fillAttr() {
        if (hint != null) ((TextInputLayout) findViewById(R.id.inputLayout)).setHint(hint);
        if (inputType > 0)
            ((TextInputEditText) findViewById(R.id.inputEditText)).setInputType(inputType);
    }

    public void setText(String text) {
        editText.setText(text);
        if (text != null && !text.isEmpty())
            setState(null, SUCCESS);
    }

    public String getText() {
        return editText.getText().toString();
    }

    @Override
    public void setOnFocusChangeListener(OnFocusChangeListener l) {
        editText.setOnFocusChangeListener(l);
    }

    public void addTextChangedListener(TextWatcher listener) {
        editText.addTextChangedListener(listener);
    }
}

File: app/src/main/java/com/hover/stax/views/StaxTextInputLayout.kt
package com.hover.stax.views

import android.content.Context
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.LayoutInflater
import com.google.android.material.textfield.TextInputEditText
import com.hover.stax.R
import com.hover.stax.databinding.StaxInputBinding


class StaxTextInputLayout(context: Context, attrs: AttributeSet) : AbstractStatefulInput(context, attrs) {

    private var hint: String? = null
    private var inputType: Int = -1
    lateinit var editText: TextInputEditText

    val binding: StaxInputBinding

    init {
        getAttrs(context, attrs)
        binding = StaxInputBinding.inflate(LayoutInflater.from(context), this, true)

        initView()
        fillAttrs()
    }

    override fun initView() {
        super.initView()
        editText = binding.inputEditText
    }

    private fun getAttrs(context: Context, attrs: AttributeSet) {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.StaxTextInputLayout, 0, 0)
        try {
            hint = a.getString(R.styleable.StaxTextInputLayout_android_hint)
            inputType = a.getInt(R.styleable.StaxTextInputLayout_android_inputType, 0)
        } finally {
            a.recycle()
        }
    }

    private fun fillAttrs() {
        hint?.let { binding.inputLayout.hint = it }
        if (inputType > 0) binding.inputEditText.inputType = inputType
    }

    fun setMutlipartText(text: String?, subtext: String?) {
        if (text.isNullOrEmpty())
            setText(subtext)
        else if (subtext.isNullOrEmpty())
            setText(text)
        else
            setText("$text ($subtext)")
    }

    fun setText(text: String?) {
        text?.let {
            editText.setText(it)

            if (it.isNotEmpty())
                setState(null, SUCCESS)
        }
    }

    fun setText(text: String?, updateState: Boolean) {
        text?.let {
            editText.setText(it)

            if (updateState && it.isNotEmpty())
                setState(null, SUCCESS)
        }
    }

    val text get() = editText.text.toString()

    override fun setOnFocusChangeListener(l: OnFocusChangeListener?) {
        editText.onFocusChangeListener = l
    }

    fun addTextChangedListener(listener: TextWatcher) = editText.addTextChangedListener(listener)
}